### PR TITLE
Add node id persistence 

### PR DESCRIPTION
**Pull request by preetapan**
Nomad agents should persist their node id, the node id is used as the server identity in raft protocol 3. 

Adds unit tests for the upgrade path to raft protocol 3, and for leader election.

### CODE REVIEW

File: nomad/leader_test.go [hunk 1]
```diff
@@ -903,6 +904,66 @@ func TestLeader_UpgradeRaftVersion(t *testing.T) {
 	}
 }
 
+func TestLeader_Reelection(t *testing.T) {
```
Comment (preetapan) [inline nomad/leader_test.go]: Still need to fix up this test to be a table driven test that tests both protocol 2 and 3 leader elections

File: command/agent/agent.go [hunk 2]
```diff
@@ -518,6 +528,66 @@ func (a *Agent) setupServer() error {
...
+func (a *Agent) setupNodeID(config *nomad.Config) error {
+	// For dev mode we have no filesystem access so just make a node ID.
+	if a.config.DevMode {
+		config.NodeID = uuid.Generate()
+		return nil
+	}
+
+	// Load saved state, if any. Since a user could edit this, we also
+	// validate it. Saved state overwrites any configured node id
+	fileID := filepath.Join(config.DataDir, "node-id")
```
Comment (schmichael) [inline command/agent/agent.go]: Should we reuse this method for the client which currently stores its ID in `client/client-id`? https://github.com/hashicorp/nomad/blob/98a403a5a692efc21a608167656e47de9f0934fe/client/client.go#L837-L842

Even if we want to share the code it's not something we need to do in this PR (or even in 0.8).

File: nomad/leader_test.go [hunk 3]
```diff
@@ -955,7 +1016,54 @@ func TestLeader_RollRaftServer(t *testing.T) {
 	TestJoin(t, s4, s1)
 	servers[1] = s4
```
Comment (dadgar) [inline nomad/leader_test.go]: Can you start with index 0 so it is easier to read

File: nomad/leader_test.go (hunk 1 above)
Comment (dadgar) [reply nomad/leader_test.go]: 👍 I believe this was your intent, but just to be explicit, lets do it as part of this PR

File: command/agent/agent.go [hunk 4]
```diff
@@ -518,6 +528,66 @@ func (a *Agent) setupServer() error {
...
+			return err
+		}
+
+		nodeID := strings.TrimSpace(string(rawID))
+		nodeID = strings.ToLower(nodeID)
+		if _, err := uuidparse.ParseUUID(nodeID); err != nil {
+			return err
+		}
+		config.NodeID = nodeID
+		savedNodeID = true
```
Comment (schmichael) [inline command/agent/agent.go]: You can just `return nil` here and avoid tracking `savedNodeID`

File: command/agent/agent.go (hunk 2 above)
Comment (jippi) [reply command/agent/agent.go]: Consul uses `${data-dir}/node-id` would be +1 on being consistent on that :)

**Pull request merged**